torch/_library/infer_schema.py

```diff
@@ -223,3 +223,23 @@ def tuple_to_list(tuple_type: typing.Type[typing.Tuple]) -> typing.Type[typing.L
         return typing.List[type_args[0]]  # type: ignore[valid-type]
     else:
         return typing.List[typing.Union[tuple(type_args)]]  # type: ignore[misc]
+
+
+def has_tensor_arg(schema: str) -> bool:
+    """
+    Given a schema string, returns True if the schema has a Tensor arg.
+    A Tensor arg is any arg with a type annotation that might involve Tensor.
+    """
+    inputs = schema.split("->")[0].split("(")[1].split(")")[0].split(",")
+    input_types = [input.strip().split(" ")[0] for input in inputs]
+    return any("Tensor" in s for s in input_types)
+
+
+def get_device_arg_id(schema: str) -> Union[int, None]:
+    """
+    Given a schema string, returns the id of the `device: torch.device` argument.
+    If it does not exist, returns None.
+    """
+    inputs = schema.split("->")[0].split("(")[1].split(")")[0].split(",")
+
+    return next((i for i, s in enumerate(inputs) if "Device device" == s.strip()), None)
```

Comment:
parsing the schema string and doing string match is a bit sketchy. Would it be possible to check the parsed schema (opoverload._schema) instead?

Reply:
These helpers (at least the first one) may be useful to put into torch._library.utils

Reply:
now using parsed schema, and moved helpers to torch._library.utils.